### PR TITLE
caf: events: module_state_event: Add a functionality to get module index

### PR DESCRIPTION
**Pull request by rakons**
This commit moves all the module identifiers into a specified section
and adds a possibility to iterate them and get its zero-based index.

Signed-off-by: Radoslaw Koppel <radoslaw.koppel@nordicsemi.no>

### CODE REVIEW

File: include/caf/events/module_state_event.h [hunk 1]
```diff
@@ -19,6 +19,36 @@
...
+
+extern const void * const MODULE_ID_LIST_START;
+extern const void * const MODULE_ID_LIST_STOP;
+
+static inline int module_count(void)
+{
+	return (&MODULE_ID_LIST_STOP - &MODULE_ID_LIST_START);
+}
+
+static inline const void * const module_id_get(int idx)
```
Comment (pdunaj) [inline include/caf/events/module_state_event.h:39]: please use `unsigned int` or `size_t` for idx

File: include/caf/events/module_state_event.h [hunk 2]
```diff
@@ -19,6 +19,36 @@
...
+#define MODULE_ID_LIST_SECTION_NAME   STRINGIFY(MODULE_ID_LIST_SECTION_PREFIX)
+#define MODULE_ID_LIST_START _CONCAT(__start_, MODULE_ID_LIST_SECTION_PREFIX)
+#define MODULE_ID_LIST_STOP  _CONCAT(__stop_,  MODULE_ID_LIST_SECTION_PREFIX)
+#define MODULE_ID_PTR_VAR_EXTERN_DEC(mname) \
+	extern const void * const MODULE_ID_PTR_VAR(mname)
+
+extern const void * const MODULE_ID_LIST_START;
+extern const void * const MODULE_ID_LIST_STOP;
+
+static inline int module_count(void)
```
Comment (pdunaj) [inline include/caf/events/module_state_event.h:34]: please return `size_t`

**Review by pdunaj — CHANGES_REQUESTED** [review]

File: include/caf/events/module_state_event.h (hunk 2 above)
Comment (pdunaj) [reply include/caf/events/module_state_event.h:34]: sorry for missing this in the original PR :(

**Comment by pdunaj** [comment]
FYI build issue related to doc is known and being fixed

File: include/caf/events/module_state_event.h (hunk 1 above)
Comment (rakons) [reply include/caf/events/module_state_event.h]: Done

File: include/caf/events/module_state_event.h (hunk 2 above)
Comment (rakons) [reply include/caf/events/module_state_event.h]: Done

**Review by pdunaj — APPROVED** [review]

**Review by MarekPieta — APPROVED** [review]

**Comment by wiminordic** [comment]
@pdunaj desktop52 devices (Gaming Mouse, Dongle and Keyboard) are failing to initialize for this PR. No errors in logs.

**Comment by pdunaj** [comment]
> @pdunaj desktop52 devices (Gaming Mouse, Dongle and Keyboard) are failing to initialize for this PR. No errors in logs.

Will have a look on a DK.

**Comment by pdunaj** [comment]
> > @pdunaj desktop52 devices (Gaming Mouse, Dongle and Keyboard) are failing to initialize for this PR. No errors in logs.
> 
> Will have a look on a DK.

I had a look and indeed it does not work correctly. It seems that value passed in the event is different to what we get from MODULE_ID(name). I have not identified yet what that is happening.

**Comment by pdunaj** [comment]
> > > @pdunaj desktop52 devices (Gaming Mouse, Dongle and Keyboard) are failing to initialize for this PR. No errors in logs.
> > 
> > 
> > Will have a look on a DK.
> 
> I had a look and indeed it does not work correctly. It seems that value passed in the event is different to what we get from MODULE_ID(name). I have not identified yet what that is happening.

The problem is at lines
```
#define MODULE_ID(mname) ({                                  \
			MODULE_ID_PTR_VAR_EXTERN_DEC(mname); \
			MODULE_ID_PTR_VAR(mname);            \
		})
```

It seems that for some reason new macros does not transform the mname correctly into the module variable name.

File: include/caf/events/module_state_event.h [hunk 3]
```diff
@@ -19,6 +19,36 @@
 extern "C" {
 #endif
 
+#define MODULE_ID_LIST_SECTION_PREFIX module_id_list
+
+#define MODULE_ID_PTR_VAR(mname) _CONCAT(__module_, MODULE)
```
Comment (pdunaj) [inline include/caf/events/module_state_event.h:24]: s/MODULE/mname

**Review by pdunaj — CHANGES_REQUESTED** [review]

File: include/caf/events/module_state_event.h (hunk 3 above)
Comment (rakons) [reply include/caf/events/module_state_event.h]: Good catch. Done.

**Review by pdunaj — APPROVED** [review]

**Pull request merged**